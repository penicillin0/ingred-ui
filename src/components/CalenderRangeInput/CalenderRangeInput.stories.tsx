import * as React from "react";
import CalenderRangeInput from "./CalenderRangeInput";
import { Story } from "@storybook/react/types-6-0";

export default {
  title: "Components/Inputs/CalenderRangeInput",
  component: CalenderRangeInput,
  args: {
    placeholder: "placeholder",
  },
};

export const Example: Story = () => <CalenderRangeInput />;
