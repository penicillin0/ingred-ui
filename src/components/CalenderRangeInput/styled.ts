import styled, { DefaultTheme } from "styled-components";

const setBorderColor = (
  theme: DefaultTheme,
  isFocused: boolean,
  isError: boolean,
) => {
  if (isError) return theme.palette.danger.main;
  else if (isFocused) return theme.palette.primary.main;
  else return theme.palette.divider;
};

export const Container = styled.div<{ isFocused: boolean; isError: boolean }>`
  display: flex;
  align-items: center;
  width: 220px;
  padding: 8px 8px;
  color: ${({ theme, isError }) =>
    isError ? theme.palette.danger.main : theme.palette.black};
  background-color: ${({ theme }) => theme.palette.background.default};
  border: 1px solid ${({ theme }) => theme.palette.divider};
  border-radius: ${({ theme }) => theme.radius}px;
  border-color: ${({ theme, isFocused, isError }) =>
    setBorderColor(theme, isFocused, isError)};
`;

export const Input = styled.input<{ letterCount: number; isError: boolean }>`
  text-align: right;
  border: none;
  font-size: 14px;
  width: ${({ letterCount }) => 8 * letterCount}px;
  color: ${({ theme, isError }) =>
    isError ? theme.palette.danger.main : theme.palette.black};
  &:focus {
    background-color: ${({ theme }) => theme.palette.primary.main};
    color: ${({ theme }) => theme.palette.white};
    outline: none;
  }
`;

export const Hyphen = styled.div`
  text-align: center;
  height: 20px;
  padding: 0px 4px;
  color: ${({ theme }) => theme.palette.gray.dark};
`;

export const Slash = styled.div`
  text-align: center;
  height: 20px;
  font-size: 12px;
`;
