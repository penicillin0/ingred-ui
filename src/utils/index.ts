export { deepmerge } from "./deepmerge";
export { createChainedFunction } from "./createChainedFunction";
export { getShadow } from "./getShadow";
export { hexToRgba } from "./hexToRgba";
export { getDuration } from "./reactTransitionGroup";
export { renderWithThemeProvider } from "./renderWithThemeProvider";
export { addScrollbarProperties } from "./scrollbar";
export { spacer, getMargin, getPadding } from "./spacer";
