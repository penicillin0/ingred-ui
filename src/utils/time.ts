import dayjs from "dayjs";
import moment from "moment";

// moment.jsで動いているreact-datesと互換性を持たせるためのメソッド
// DatePickerとDateRangePickerのリニューアルが完了したら除去する
export function dayjsToMoment(date: dayjs.Dayjs | null): moment.Moment | null {
  if (!date) return null;
  const dateString = date.format();
  return moment(dateString);
}

export function momentToDayjs(date: moment.Moment | null): dayjs.Dayjs | null {
  if (!date) return null;
  const dateString = date.format();
  return dayjs(dateString);
}
