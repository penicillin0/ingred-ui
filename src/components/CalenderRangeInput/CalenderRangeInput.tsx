import React, { useState, useRef } from "react";
import * as Styled from "./styled";
import dayjs from "dayjs";

const CalenderRangeInput: React.FC = () => {
  const today = dayjs();
  const [startYear, setStartYear] = useState<number>(today.year());
  const [startMonth, setStartMonth] = useState<number>(today.month() + 1);
  const [startDate, setStartDate] = useState<number>(today.date());
  const [endYear, setEndYear] = useState<number>(today.year());
  const [endMonth, setEndMonth] = useState<number>(today.month() + 1);
  const [endDate, setEndDate] = useState<number>(today.date());
  const [isFocused, setIsFocused] = useState<boolean>(false);
  const [isError] = useState<boolean>(false);
  const startYearInput = useRef<HTMLInputElement>(null);
  const startMonthInput = useRef<HTMLInputElement>(null);
  const startDateInput = useRef<HTMLInputElement>(null);
  const endYearInput = useRef<HTMLInputElement>(null);
  const endMonthInput = useRef<HTMLInputElement>(null);
  const endDateInput = useRef<HTMLInputElement>(null);

  const handleOnChange =
    (datePart: string) => (event: React.ChangeEvent<HTMLInputElement>) => {
      if (datePart === "startYear") setStartYear(Number(event.target.value));
      if (datePart === "startMonth") setStartMonth(Number(event.target.value));
      if (datePart === "startDate") setStartDate(Number(event.target.value));
      if (datePart === "endYear") setEndYear(Number(event.target.value));
      if (datePart === "endMonth") setEndMonth(Number(event.target.value));
      if (datePart === "endDate") setEndDate(Number(event.target.value));
    };

  const handleOnKeyPress =
    (datePart: string) => (event: React.KeyboardEvent<HTMLInputElement>) => {
      if (event.key === "Enter" || event.key === "ArrowRight") {
        if (datePart === "startYear") {
          if (!startYearInput.current || !startYearInput.current.selectionStart)
            return;
          if (startYearInput.current.selectionStart === 4)
            startMonthInput.current?.focus();
        }
        if (datePart === "startMonth") {
          if (
            !startMonthInput.current ||
            !startMonthInput.current.selectionStart
          )
            return;
          if (startMonthInput.current.selectionStart === 2)
            startDateInput.current?.focus();
        }
        if (datePart === "startDate") {
          if (!startDateInput.current || !startDateInput.current.selectionStart)
            return;
          if (startDateInput.current.selectionStart === 2)
            endYearInput.current?.focus();
        }
        if (datePart === "endYear") {
          if (!endYearInput.current || !endYearInput.current.selectionStart)
            return;
          if (endYearInput.current.selectionStart === 4)
            endMonthInput.current?.focus();
        }
        if (datePart === "endMonth") {
          if (!endMonthInput.current || !endMonthInput.current.selectionStart)
            return;
          if (endMonthInput.current.selectionStart === 2)
            endDateInput.current?.focus();
        }
      }
      if (event.key === "ArrowLeft") {
        if (datePart === "startMonth") {
          if (
            !startMonthInput.current ||
            !startMonthInput.current.selectionStart
          )
            return;
          if (startMonthInput.current.selectionStart === 1)
            startYearInput.current?.focus();
        }
        if (datePart === "startDate") {
          if (!startDateInput.current || !startDateInput.current.selectionStart)
            return;
          if (startDateInput.current.selectionStart === 1)
            startMonthInput.current?.focus();
        }
        if (datePart === "endYear") {
          if (!endYearInput.current || !endYearInput.current.selectionStart)
            return;
          if (endYearInput.current.selectionStart === 1)
            startDateInput.current?.focus();
        }
        if (datePart === "endMonth") {
          if (!endMonthInput.current || !endMonthInput.current.selectionStart)
            return;
          if (endMonthInput.current.selectionStart === 1)
            endYearInput.current?.focus();
        }
        if (datePart === "endDate") {
          if (!endDateInput.current || !endDateInput.current.selectionStart)
            return;
          if (endDateInput.current.selectionStart === 1)
            endMonthInput.current?.focus();
        }
      }
      if (event.key === "ArrowUp") {
        if (datePart === "startYear") setStartYear(startYear + 1);
        if (datePart === "startMonth" && startMonth < 12)
          setStartMonth(startMonth + 1);
        if (datePart === "startDate" && startDate < 31)
          setStartDate(startDate + 1);
        if (datePart === "endYear") setEndYear(endYear + 1);
        if (datePart === "endMonth" && endMonth < 12) setEndMonth(endMonth + 1);
        if (datePart === "endDate" && endDate < 31) setEndDate(endDate + 1);
      }

      if (event.key === "ArrowDown") {
        if (datePart === "startYear") setStartYear(startYear - 1);
        if (datePart === "startMonth" && startMonth > 1)
          setStartMonth(startMonth - 1);
        if (datePart === "startDate" && startDate > 1)
          setStartDate(startDate - 1);
        if (datePart === "endYear") setEndYear(endYear - 1);
        if (datePart === "endMonth" && endMonth > 1) setEndMonth(endMonth - 1);
        if (datePart === "endDate" && endDate > 1) setEndDate(endDate - 1);
      }
    };

  const handleInputNumberOnly = (
    event: React.ChangeEvent<HTMLInputElement>,
  ) => {
    event.target.value = event.target.value.replace(/[^0-9]+/i, "");
  };

  const handleOnFocus = () => {
    setIsFocused(true);
  };
  const handleOnBlur = () => {
    if (
      startYearInput.current !== document.activeElement &&
      startMonthInput.current !== document.activeElement &&
      startDateInput.current !== document.activeElement &&
      endYearInput.current !== document.activeElement &&
      endMonthInput.current !== document.activeElement &&
      endDateInput.current !== document.activeElement
    ) {
      setIsFocused(false);
    }
  };

  return (
    <Styled.Container isFocused={isFocused} isError={isError}>
      <Styled.Input
        ref={startYearInput}
        isError={isError}
        letterCount={4}
        type={"text"}
        aria-label="startYear"
        value={startYear}
        onChange={handleOnChange("startYear")}
        onKeyDown={handleOnKeyPress("startYear")}
        onInput={handleInputNumberOnly}
        onFocus={handleOnFocus}
        onBlur={handleOnBlur}
      />
      <Styled.Slash>/</Styled.Slash>
      <Styled.Input
        ref={startMonthInput}
        isError={isError}
        letterCount={2}
        type={"text"}
        aria-label="startMonth"
        value={("00" + startMonth).slice(-2)}
        onChange={handleOnChange("startMonth")}
        onKeyDown={handleOnKeyPress("startMonth")}
        onInput={handleInputNumberOnly}
        onFocus={handleOnFocus}
        onBlur={handleOnBlur}
      />
      <Styled.Slash>/</Styled.Slash>
      <Styled.Input
        ref={startDateInput}
        isError={isError}
        letterCount={2}
        type={"text"}
        aria-label="startDate"
        value={("00" + startDate).slice(-2)}
        onChange={handleOnChange("startDate")}
        onKeyDown={handleOnKeyPress("startDate")}
        onInput={handleInputNumberOnly}
        onFocus={handleOnFocus}
        onBlur={handleOnBlur}
      />
      <Styled.Hyphen>-</Styled.Hyphen>
      <Styled.Input
        ref={endYearInput}
        isError={isError}
        letterCount={4}
        type={"text"}
        aria-label="endYear"
        value={endYear}
        onChange={handleOnChange("endYear")}
        onKeyDown={handleOnKeyPress("endYear")}
        onInput={handleInputNumberOnly}
        onFocus={handleOnFocus}
        onBlur={handleOnBlur}
      />
      <Styled.Slash>/</Styled.Slash>
      <Styled.Input
        ref={endMonthInput}
        isError={isError}
        letterCount={2}
        type={"text"}
        aria-label="endMonth"
        value={("00" + endMonth).slice(-2)}
        onChange={handleOnChange("endMonth")}
        onKeyDown={handleOnKeyPress("endMonth")}
        onInput={handleInputNumberOnly}
        onFocus={handleOnFocus}
        onBlur={handleOnBlur}
      />
      <Styled.Slash>/</Styled.Slash>
      <Styled.Input
        ref={endDateInput}
        isError={isError}
        letterCount={2}
        type={"text"}
        aria-label="endDate"
        value={("00" + endDate).slice(-2)}
        onChange={handleOnChange("endDate")}
        onKeyDown={handleOnKeyPress("endDate")}
        onInput={handleInputNumberOnly}
        onFocus={handleOnFocus}
        onBlur={handleOnBlur}
      />
    </Styled.Container>
  );
};

export default CalenderRangeInput;
