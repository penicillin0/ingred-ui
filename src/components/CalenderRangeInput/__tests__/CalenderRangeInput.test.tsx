import * as React from "react";
import "@testing-library/jest-dom/extend-expect";
import { cleanup } from "@testing-library/react";
import CalenderRangeInput from "..";
import { renderWithThemeProvider } from "../../../utils/renderWithThemeProvider";

describe("CalenderRangeInput component testing", () => {
  afterEach(cleanup);

  test("CalenderRangeInput", () => {
    const { asFragment } = renderWithThemeProvider(<CalenderRangeInput />);
    expect(asFragment()).toMatchSnapshot();
  });
});
