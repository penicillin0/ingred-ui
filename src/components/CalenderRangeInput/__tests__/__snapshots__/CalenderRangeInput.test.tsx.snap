// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalenderRangeInput component testing CalenderRangeInput 1`] = `
<DocumentFragment>
  <div
    class="sc-bczRLJ fIyvYk"
  >
    <input
      aria-label="startYear"
      class="sc-gsnTZi rexAi"
      type="text"
      value="2022"
    />
    /
    <input
      aria-label="startMonth"
      class="sc-gsnTZi kPbdrI"
      type="text"
      value="12"
    />
    /
    <input
      aria-label="startDate"
      class="sc-gsnTZi kPbdrI"
      type="text"
      value="12"
    />
    <div
      class="sc-dkzDqf ehnfnL"
    >
      -
    </div>
    <input
      aria-label="endYear"
      class="sc-gsnTZi rexAi"
      type="text"
      value="2022"
    />
    /
    <input
      aria-label="endMonth"
      class="sc-gsnTZi kPbdrI"
      type="text"
      value="12"
    />
    /
    <input
      aria-label="endDate"
      class="sc-gsnTZi kPbdrI"
      type="text"
      value="12"
    />
  </div>
</DocumentFragment>
`;
